coderd/parameters.go

```diff
@@ -0,0 +1,244 @@
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	owner, err := api.getWorkspaceOwnerData(ctx, user, templateVersion.OrganizationID)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error fetching workspace owner.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	input := preview.Input{
+		PlanJSON:        plan,
+		ParameterValues: map[string]string{},
+		Owner:           owner,
+	}
+
+	conn, err := websocket.Accept(rw, r, nil)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusUpgradeRequired, codersdk.Response{
+			Message: "Failed to accept WebSocket.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+	stream := wsjson.NewStream[codersdk.DynamicParametersRequest, codersdk.DynamicParametersResponse](
+		conn,
+		websocket.MessageText,
+		websocket.MessageText,
+		api.Logger,
+	)
+
+	// Send an initial form state, computed without any user input.
+	result, diagnostics := preview.Preview(ctx, input, fs)
+	response := codersdk.DynamicParametersResponse{
+		ID:          -1,
+		Diagnostics: previewtypes.Diagnostics(diagnostics),
+	}
+	if result != nil {
+		response.Parameters = result.Parameters
+	}
+	err = stream.Send(response)
+	if err != nil {
+		stream.Drop()
+		return
+	}
+
+	// As the user types into the form, reprocess the state using their input,
+	// and respond with updates.
+	updates := stream.Chan()
+	for {
+		select {
+		case <-ctx.Done():
+			stream.Close(websocket.StatusGoingAway)
+			return
+		case update, ok := <-updates:
+			if !ok {
+				// The connection has been closed, so there is no one to write to
+				return
+			}
+			input.ParameterValues = update.Inputs
+			result, diagnostics := preview.Preview(ctx, input, fs)
+			response := codersdk.DynamicParametersResponse{
+				ID:          update.ID,
+				Diagnostics: previewtypes.Diagnostics(diagnostics),
+			}
+			if result != nil {
+				response.Parameters = result.Parameters
+			}
+			err = stream.Send(response)
+			if err != nil {
+				stream.Drop()
+				return
+			}
+		}
+	}
+}
+
+func (api *API) getWorkspaceOwnerData(
+	ctx context.Context,
+	user database.User,
+	organizationID uuid.UUID,
+) (previewtypes.WorkspaceOwner, error) {
+	var g errgroup.Group
+
+	var ownerRoles []previewtypes.WorkspaceOwnerRBACRole
+	g.Go(func() error {
+		orgRoles, err := api.Database.GetOrganizationMemberRoles(ctx, database.GetOrganizationMemberRolesParams{
+			OrganizationID: organizationID,
+			UserID:         user.ID,
+		})
+		if err != nil {
+			return err
+		}
+		ownerRoles = make([]previewtypes.WorkspaceOwnerRBACRole, 0, len(user.RBACRoles)+len(orgRoles))
+		for _, it := range user.RBACRoles {
+			ownerRoles = append(ownerRoles, previewtypes.WorkspaceOwnerRBACRole{
+				Name: it,
+			})
+		}
+		for _, it := range orgRoles {
+			ownerRoles = append(ownerRoles, previewtypes.WorkspaceOwnerRBACRole{
+				Name:  it,
+				OrgID: organizationID,
+			})
+		}
+		return nil
+	})
```

Comment:
This only returns their org roles, but we actually want to provide all roles.

We should match how apikey does it: https://github.com/coder/coder/blob/2b4c609b5c01b514cce6c8d265a9d62f8166e480/coderd/httpmw/apikey.go#L447-L477

Since we only need the role names, you can see how the provisioner grabs these roles:

https://github.com/coder/coder/blob/2b4c609b5c01b514cce6c8d265a9d62f8166e480/coderd/provisionerdserver/provisionerdserver.go#L598-L609

Ideally the code that populates in the provisioner is reused for the dynamic params.
https://github.com/coder/coder/blob/2b4c609b5c01b514cce6c8d265a9d62f8166e480/coderd/provisionerdserver/provisionerdserver.go#L619-L640